File: src/main/java/fi/aalto/cs/apluscourses/ui/FileSaveView.java
package fi.aalto.cs.apluscourses.ui;

import static fi.aalto.cs.apluscourses.utils.PluginResourceBundle.getAndReplaceText;
import static fi.aalto.cs.apluscourses.utils.PluginResourceBundle.getText;

import com.intellij.openapi.fileChooser.FileChooserFactory;
import com.intellij.openapi.fileChooser.FileSaverDescriptor;
import com.intellij.openapi.fileChooser.FileSaverDialog;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFileWrapper;
import fi.aalto.cs.apluscourses.presentation.FileSaveViewModel;
import java.nio.file.InvalidPathException;
import javax.swing.JOptionPane;
import org.jetbrains.annotations.NotNull;

public class FileSaveView implements Dialog {

  @NotNull
  private final FileSaveViewModel viewModel;

  @NotNull
  private final Project project;

  public FileSaveView(@NotNull FileSaveViewModel viewModel,
                      @NotNull Project project) {
    this.viewModel = viewModel;
    this.project = project;
  }

  @Override
  public boolean showAndGet() {
    FileSaverDescriptor descriptor
        = new FileSaverDescriptor(viewModel.getTitle(), viewModel.getDescription());

    while (true) {
      FileSaverDialog dialog = FileChooserFactory
          .getInstance()
          .createSaveFileDialog(descriptor, project);
      VirtualFileWrapper file = dialog.save(
          viewModel.getDefaultDirectory(), viewModel.getDefaultName());

      if (file != null) {
        try {
          viewModel.setPath(file.getFile().toPath());
          return true;
        } catch (InvalidPathException e) {
          JOptionPane.showMessageDialog(null,
              getAndReplaceText("ui.exportModule.invalidPath.message", e.getReason()),
              getText("ui.exportModule.invalidPath.title"),
              JOptionPane.ERROR_MESSAGE);
          continue;
        }
      }

      return false;
    }
  }
}

File: src/main/kotlin/fi/aalto/cs/apluscourses/ui/temp/FileSaveView.kt
package fi.aalto.cs.apluscourses.ui.temp

import com.intellij.openapi.fileChooser.FileChooserFactory
import com.intellij.openapi.fileChooser.FileSaverDescriptor
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import fi.aalto.cs.apluscourses.MyBundle.message
import fi.aalto.cs.apluscourses.ui.temp.presentation.FileSaveViewModel
import java.nio.file.InvalidPathException

class FileSaveView(
    viewModel: FileSaveViewModel,
    project: Project
) {
    private val viewModel: FileSaveViewModel

    private val project: Project

    init {
        this.viewModel = viewModel
        this.project = project
    }

    fun showAndGet(): Boolean {
        val descriptor = FileSaverDescriptor(viewModel.getTitle(), viewModel.getDescription())

        while (true) {
            val dialog = FileChooserFactory
                .getInstance()
                .createSaveFileDialog(descriptor, project)
            val file = dialog.save(
                viewModel.getDefaultDirectory(), viewModel.getDefaultName()
            )

            if (file != null) {
                try {
                    viewModel.setPath(file.file.toPath())
                    return true
                } catch (e: InvalidPathException) {
                    Messages.showErrorDialog(
                        message("ui.exportModule.invalidPath.message", e.reason),
                        message("ui.exportModule.invalidPath.title")
                    )
                    continue
                }
            }

            return false
        }
    }
}
